crates/rpc/rpc/src/eth/pubsub.rs

```diff
@@ -1,63 +1,128 @@
     }
 }
 
 impl<Client, Pool> EthPubSubApiServer for EthPubSub<Client, Pool>
 where
-    Client: BlockProvider + 'static,
+    Client: BlockProvider + Clone + 'static,
     Pool: TransactionPool + 'static,
 {
     fn subscribe(
         &self,
         mut sink: SubscriptionSink,
-        _kind: Kind,
-        _params: Option<Params>,
+        kind: SubscriptionKind,
+        params: Option<Params>,
     ) -> SubscriptionResult {
         sink.accept()?;
-        todo!()
+
+        let pubsub = self.inner.clone();
+        self.subscription_task_spawner.spawn(Box::pin(async move {
+            handle_accepted(pubsub, sink, kind, params).await;
+        }));
+
+        Ok(())
     }
 }
 
 /// The actual handler for and accepted [`EthPubSub::subscribe`] call.
 async fn handle_accepted<Client, Pool>(
-    _pool: Pool,
-    _client: Arc<Client>,
-    _accepted_sink: SubscriptionSink,
-    _kind: Kind,
-    _params: Option<Params>,
-) {
+    pubsub: EthPubSubInner<Client, Pool>,
+    mut accepted_sink: SubscriptionSink,
+    kind: SubscriptionKind,
+    params: Option<Params>,
+) where
+    Client: BlockProvider + 'static,
+    Pool: TransactionPool + 'static,
+{
+    // if no params are provided, used default filter params
+    let _params = match params {
+        Some(Params::Logs(filter)) => FilteredParams::new(Some(*filter)),
+        _ => FilteredParams::default(),
+    };
+
+    match kind {
+        SubscriptionKind::NewHeads => {
+            // TODO subscribe new blocks -> convert
+        }
+        SubscriptionKind::Logs => {
+            // TODO subscribe new blocks -> fetch logs via bloom
+        }
+        SubscriptionKind::NewPendingTransactions => {
+            let stream = pubsub
+                .into_pending_transaction_stream()
+                .map(EthSubscriptionResult::TransactionHash);
+            accepted_sink.pipe_from_stream(stream).await;
+        }
+        SubscriptionKind::Syncing => {
+            // TODO subscribe new blocks -> read is_syncing from network
+        }
+    }
+}
+
+impl<Client, Pool> std::fmt::Debug for EthPubSub<Client, Pool> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("EthPubSub").finish_non_exhaustive()
+    }
 }
 
 /// Container type `EthPubSub`
-#[derive(Debug)]
+#[derive(Clone)]
 struct EthPubSubInner<Client, Pool> {
     /// The transaction pool.
     pool: Pool,
     /// The client that can interact with the chain.
-    client: Arc<Client>,
-    // TODO needs spawn access
+    client: Client,
+}
+
+// == impl EthPubSubInner ===
+
+impl<Client, Pool> EthPubSubInner<Client, Pool>
+where
+    Client: BlockProvider + 'static,
+    Pool: TransactionPool + 'static,
+{
+    /// Returns a stream that yields all transactions emitted by the txpool.
+    fn into_pending_transaction_stream(self) -> impl Stream<Item = TxHash> {
+        ReceiverStream::new(self.pool.pending_transactions_listener())
+    }
 }
```

Comment:
as a non standard feature, i think we should support full transactions exposed here, i.e. not just their hash. maybe we can do it later but somethign to keep in mind. maybe there's an EIP for this actually.